ratelimit: refactor rate limit code and move to extensions

ramaraochavali opened the pull request:
*Description*: This is part of a multi step process to implement https://github.com/envoyproxy/envoy/issues/4669. This PR moves the rate limiting code from core to extensions, enabling the other steps.
*Risk Level*: Low, it is just a package refactor not a functional change.
*Testing*:  Existing automated tests
*Docs Changes*: N/A
*Release Notes*: N/A

ramaraochavali commented:
@mattklein123  As discussed, here is the first PR to implement #4669 . There are some minor deviations from what we have discussed there.

- I could not put the rate limit client in to `Singleton` so that all filters can use a single client, because of the following reasons.
  - The client has to be created on the main thread because of how the `Singleton` manager works. So I had to create client once when Listener Manager instantiates the filters in the `createFilterFactoryFromProtoTyped` method. So here we have to check in the singleton manager first, if it is not available create one and hold the client in the filter. This is not a problem, it actually works.
  - Once the client is stored in the filter, the `GrpcClientImpl` also has `AsyncRequest` . This seems to be a problem sharing this across threads. I ran in to this assert https://github.com/envoyproxy/envoy/blob/master/source/common/event/dispatcher_impl.cc#L45 because the `AsyncRequest` is `DeferredDeletable`. Not sure how to make this share across threads without safely without adding too much of complexity. Probably we should add `AsyncRequest` to TLS?
- Also could not add  the `RateLimitServiceConfig` to `Singleton`. This is because, we have to add it to `Singleton` here https://github.com/envoyproxy/envoy/blob/master/source/server/configuration_impl.cc#L83. But we have to access it in config.cc of filter. I was not sure how to access it there using the `SINGLETON_MANAGER_REGISTERED_NAME` macro. I had to hard code name. So I have added a method `rateLimitServiceConfig` in `FactoryContext` to get it, which I think is equivalent to `Singleton`.
Can you PTAL and suggest if there are any better ways of doing this? otherwise it is good for review.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.cc:
```diff
@@ -1,20 +1,35 @@
 #include <vector>
 
 #include "envoy/api/v2/ratelimit/ratelimit.pb.h"
+#include "envoy/registry/registry.h"
```
This is left over from my singleton experiments. Will cleanup if we decide not to use singleton.

mattklein123 commented on include/envoy/ratelimit/BUILD:
```diff
@@ -13,6 +13,9 @@ envoy_cc_library(
     hdrs = ["ratelimit.h"],
     external_deps = ["abseil_optional"],
     deps = [
+        "//include/envoy/registry",
```
This header file should be moved entirely into extensions

ramaraochavali replied:
I could not completely move this to extensions  because, the router implementation https://github.com/envoyproxy/envoy/blob/master/source/common/router/router_ratelimit.cc uses `Descriptor` and `DescriptorEntry`  classes.  So I kept those classes here.

mattklein123 commented on include/envoy/ratelimit/ratelimit.h:
```diff
@@ -6,39 +6,14 @@
 #include <vector>
 
 #include "envoy/common/pure.h"
+#include "envoy/config/bootstrap/v2/bootstrap.pb.h"
```
Why is bootstrap needed here? Anything in here should be an implementation detail I think.

ramaraochavali replied:
Ok. I think I can move the class `RateLimitServiceConfig` that I added to `configuration_impl.h`.

mattklein123 commented on include/envoy/server/instance.h:
```diff
@@ -213,6 +206,11 @@ class Instance {
    */
   virtual std::chrono::milliseconds statsFlushInterval() const PURE;
+
+  /**
```
Why is this still here? All of this should be gone from the public interfaces.

ramaraochavali replied:
This is because of the second singleton problem I am talking about. If there is a way to get singleton name/it is ok to hard code the singleton name, I can remove this.

ramaraochavali replied:
I have removed from this and other public interfaces.

ramaraochavali commented:
@mattklein123 I have added the TLS version of rate limit client and moved it to singleton manager. PTAL. For registering, ratelimit service config with singleton, I have a question - if that is ok I can move it there.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.cc:
```diff
@@ -0,0 +1,146 @@
+  // TODO(ramaraochavali): figure out a way to get the singleton name here.
+  Envoy::RateLimit::RateLimitServiceConfigPtr ratelimit_config =
+      context.singletonManager().getTyped<Envoy::RateLimit::RateLimitServiceConfig>(
+          "ratelimit_config_singleton", [] { return nullptr; });
```
@mattklein123  This is the second problem with singleton I am taking about. I have registered the `ratelimit_config` in ConfigurationImpl.cc where we setup the config and added it to singleton there. Now in the extensions, I want to get the already registered singleton object. I could not find a way to get it - I have to hard the singleton name here with a dummy lamda that should never execute if rate limit service config is there. Am I using it correctly? If there is a way to get the rate limit service config set in singleton without hard coding the name - I can remove the method that I added in FactoryContext to get the rateLimitServiceConfig that will remove from Instance.h and Server.h .
If it is ok, to move forward like this also I can remove the rateLimitServiceConfig method from FactoryContext and other places. LMK.

ramaraochavali replied:
Yeah. That is a good option. Instead of creating client from core, I think we have to just register the rate limit config by calling in to extensions. Once the rate limit config is registered, the filter config code can build client using the registered rate limit config because any way the client has to be built there. Also when we introduce rate limit service config in filter, there will be minimal changes because we can also look at the config in the filter and build it, validate it etc. I have made changes like that. PTAL.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.h:
```diff
@@ -0,0 +1,169 @@
+  ThreadLocal::SlotAllocator& tls_;
+};
+
+class NullClientImpl : public Client {
```
Yes. I will add comments.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.cc:
```diff
@@ -0,0 +1,148 @@
+    : tls_(tls) {
+  envoy::api::v2::core::GrpcService grpc_service;
+  grpc_service.MergeFrom(config.grpc_service());
+  // TODO(htuch): cluster_name is deprecated, remove after 1.6.0.
```
Not just the code but api also needs to be removed. Not sure if there is an impact on tests.  ithttps://github.com/envoyproxy/envoy/blob/b6ff832b9159fad4aef3e93d6243753f9427494c/api/envoy/config/ratelimit/v2/rls.proto#L23
Probably separate PR so that it is clean? I can do that once this is done.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_registration.cc:
```diff
@@ -0,0 +1,34 @@
+RateLimitServiceConfigPtr rateLimitConfig(Server::Configuration::FactoryContext& context) {
+  RateLimitServiceConfigPtr ratelimit_config =
+      context.singletonManager().getTyped<RateLimitServiceConfig>(
+          SINGLETON_MANAGER_REGISTERED_NAME(ratelimit_config), [] { return nullptr; });
```
I think this can not be avoided. Even if we build client directly from core code, the filter config code would need a way to get the built client so there also same problem would come. I felt this is much cleaner and aligns with future code. I can add an ASSERT here to indicate that lambda should never be called.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.cc:
```diff
@@ -0,0 +1,145 @@
+namespace Common {
+namespace RateLimit {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
```
Currently also it does not allow registering twice - Here I am registering client. I can move that to registration file.

ramaraochavali replied:
On a second thought, I think it is better to leave the singleton registration of client here it self because, once we fully migrate to the new model and remove rate limit service from bootstrap, we can just remove the whole registration file. I will documentation to indicate the same.

mattklein123 replied:
Sorry I still do not understand why we have to singleton register the config at all. I think we should only be singleton registering the client:
1) Bootstrap -> Create client (or null) and register it
2) Filter, get singleton client

In the future, we can add client config, and then in 2) we can also attempt to register, and then we can remove 1. What am I missing?

ramaraochavali replied:
We can do what you said above. But we can only register the client once. So if we add the client config, we can not register from multiple places (bootstrap and filter). We will have to remove at one place - we are not sure at which place we can remove when both co-exist during the deprecation period.  The advantage of registering config is we can validate during the deprecation period that one supplied in bootstrap and client config is same and register client at only one place (that is from filter code as I am doing today). If the validation fails, we can fail fast. If we do not store config in singleton, I do not know how it we can work during the deprecation period - if core code already registers client and we can not register again with the client config in filters.
So leaving the registration of client in the filter code will allow us to validate and register it with right config (either bootstrap or client)?
Does it make sense?

mattklein123 replied:
The point of the singleton manager is that you can attempt to register from multiple places, but it will only get registered once. So basically you have common code that looks like this:
1) See if singleton exists
2) If so, check if config matches, if not fail config
3) If not, create singleton

This should work fine.

ramaraochavali replied:
When I attempt to register the client from filter code with filter config, how can I get the bootstrap config to validate in 2 above? Are you suggesting to hold the config in the GrpcClientImpl (global rate limit client)? I am probably missing that part.

mattklein123 replied:
Yes, I would just store the proto config used to create the client, and then compare it.

ramaraochavali replied:
Ok. Now it make sense. I will make that change.

ramaraochavali replied:
I moved the whole client creation/registration to ratelimit_registration so that singletons are maintained at single place.

mattklein123 replied:
OK that's fine. I think we can clean this up once we deprecate the bootstrap config, and then we can have a single client for the entire server which would be nice.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_registration.cc:
```diff
@@ -0,0 +1,34 @@
+namespace Common {
+namespace RateLimit {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
```
Ultimately, we just want the client in the singleton not the config. Config was added in the singleton as an interim step, which will be removed once we remove rate limit config from bootstrap. Then this whole file will go away. So it will be a least disruptive change. The validation will be added in the filter config code. Will documentation around that.

mattklein123 replied:
Sorry I don't follow. Why can't you store the client in the singleton manager in this change? It should be fine and what you want in the future anyway.

ramaraochavali replied:
I am storing client in singleton manager already in this change. I am just saying i am also storing config in the interim.

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.cc:
```diff
@@ -0,0 +1,145 @@
+
+ClientPtr GrpcFactoryImpl::create(const absl::optional<std::chrono::milliseconds>& timeout,
+                                  Server::Configuration::FactoryContext& context) {
+  return context.singletonManager().getTyped<GrpcClientImpl>(
```
@mattklein123 Here is where I am storing client in singleton manager and this will remain as is in future also. We do not have to change this part. That is why I am not moving this to ratelimit_registration.cc. My plan is that ratelimit_registration will completely go away when we finally delete the rate limit config from bootstrap.

jmarantz commented:
/sub

ramaraochavali commented on source/extensions/filters/common/ratelimit/ratelimit_impl.h:
```diff
@@ -0,0 +1,169 @@
+  void onCreateInitialMetadata(Http::HeaderMap&) override {}
+  void onSuccess(std::unique_ptr<envoy::service::ratelimit::v2::RateLimitResponse>&& response,
+                 Tracing::Span& span) override {
+    tls_slot_->getTyped<GrpcTlsClientImpl>().onSuccess(std::move(response), span);
```
No problem. unwinding is easy :-). Reverted the TLS part altogether. PTAL. I think in the order of priority this is what we should as next steps once this is merged.
- Remove the cluster_name deprecated stuff
- Implement the filter config for RLS and make it work like External Autz and deprecate bootstrap
- See if code reuse possible between external autz and this refactor it
- Make client singleton supporting all of the things you said (make client work for concurrent requests as part of this or before this)

mattklein123 commented on source/extensions/filters/common/ratelimit/ratelimit_registration.cc:
```diff
@@ -0,0 +1,61 @@
+        // This should never happen. We expect config to be registered to singleton, during
+        // configuration processing in the core at start up.
+        NOT_REACHED_GCOVR_EXCL_LINE;
+        return nullptr;
```
I think you can delete this line

ramaraochavali replied:
I tried deleting that line. If I delete this line, compiler is complaining. So left it.

mattklein123 commented on source/extensions/filters/common/ratelimit/ratelimit_registration.cc:
```diff
@@ -0,0 +1,61 @@
+                          RateLimitServiceConfigPtr ratelimit_config, const uint32_t timeout_ms) {
+  // When we introduce rate limit service config in filters, we should validate here that it
+  // matches with bootstrap.
+  ASSERT(ratelimit_config != nullptr);
```
You can delete this assert, it will crash obviously below.

mattklein123 commented on source/extensions/filters/common/ratelimit/ratelimit_registration.cc:
```diff
@@ -0,0 +1,61 @@
+  // When we introduce rate limit service config in filters, we should validate here that it
+  // matches with bootstrap.
+  ASSERT(ratelimit_config != nullptr);
+  ClientFactoryPtr client_factory;
```
Is it possible to actually store the factory in the singleton manager vs. the config? I think that should work? This would get us a little closer to where we want, and would also reduce an allocation that you have now added for every HTTP request?

ramaraochavali replied:
I think we can store the factory in the singleton manager. Only thing then is the factory needs to hold the config with which it is created, so that we can add the validation in future. We can add the logic to hold it when we implement validation. So now, removed config and added factory to singleton. PTAL.

ramaraochavali commented:
/retest

ramaraochavali commented:
@mattklein123 factory in and config out of singleton :-). PTAL.

mattklein123 commented on source/extensions/filters/common/ratelimit/ratelimit_registration.cc:
```diff
@@ -0,0 +1,56 @@
+      });
+}
+
+ClientPtr rateLimitClient(ClientFactoryPtr client_factory, const uint32_t timeout_ms) {
```
This function doesn't add any value right now, can you remove and just make the call below inline at the call sites. We can add a helper function back in a future change.

ramaraochavali commented:
@mattklein123 removed the method. PTAL.
